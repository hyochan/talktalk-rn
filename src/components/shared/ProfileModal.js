import LinearGradient from 'react-native-linear-gradient';
import React, {
  Component,
  useState,
  useImperativeHandle,
  forwardRef,
} from 'react';
import {
  ActivityIndicator,
  TouchableOpacity,
  StyleSheet,
  Image,
  Text,
  View,
  ViewStyle,
  ImageStyle,
  TextStyle,
  ImageSourcePropType,
} from 'react-native';
import Modal from 'react-native-modalbox';

import Icon5 from 'react-native-vector-icons/FontAwesome5';
import styled from 'styled-components/native';

import { User } from '../../models/User';
import { IC_MASK } from '../../utils/Icons';
import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

const StyledWrapper = styled.View`
  height: 320;
  margin-horizontal: 20;
  aligin-self: stretch;
  align-items: center;
  justify-content: space-between;
`;

const StyledView = styled.View`
  marginTop: 40;
`;

const StyledViewBtns = styled.View`
  height: 80;
  align-self: stretch;
  background-color: white;

  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledViewBtnDivider = styled.View`
  width: 1;
  height: 80;
  background-color: ${colors.paleGray};
`;

const StyledTextDisplayName = styled.Text`
  font-size: 24;
  color: white;
  font-weight: bold;
  margin-top: 32;
  align-self: center;
`;

const StyledTextStatusMsg = styled.Text`
  font-size: 12;
  color: white;
  margin-top: 8;
  align-self: center;
`;

const StyledTextBtn = styled.Text`
  color: ${colors.dodgerBlue};
  font-size: 16;
`;

const StyledTextFriendAdded = styled.Text`
  color: white;
  font-size: 12;
  backround-color: ${colors.dusk};
  padding: 4;
`;

const StyledTextFriendAlreadyAdded = styled.Text`
  color: red;
  font-size: 12;
  background-color: ${colors.cloudyBlue};
  padding: 4;
`;

type Props = {
  onChatPressed?: () => void;
};

type Ref = {
  open: () => void;
  close: () => void;
  addFriend: () => void;
  deleteFriend: () => void;
};

type Styles = {
  viewBtn: ViewStyle,
};

const styles = {
  wrapper: {
    backgroundColor: 'transparent',
    alignSelf: 'stretch',
    height: 320,
    width: '90%',

    alignContent: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  viewBtn: {
    width: '50%',
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
};

function Shared(props: Props, ref: Ref) {
  let modal: any;
  const [showAddBtn, setShowAddBtn] = useState(true);
  const [isAdding, setIsAdding] = useState(false);
  const [isFriendAdded, setIsFriendAdded] = useState(false);
  const [isFriendAlreadyAdded, setIsFriendAlreadyAdded] = useState(false);
  const [user, setUser] = useState(new User());

  const open = () => {
    setIsFriendAdded(false);
    setIsFriendAlreadyAdded(false);
    if (modal) {
      modal.open();
    }
  };

  const close = () => {
    if (modal) {
      modal.close();
    }
  };

  const addFriend = () => {

  };

  const deleteFriend = () => {
    if (modal) {
      modal.close();
    }
  };

  useImperativeHandle(ref, () => ({
    open,
    close,
    setUser: (user: User) => {
      setUser(user);
    },
    showAddBtn: (flag: boolean) => {
      setShowAddBtn(flag);
    },
  }));

  return (
    <Modal
      ref={(v) => modal = v}
      backdropOpacity={0.075}
      entry={'top'}
      position={'center'}
      style={styles.wrapper}
    >
      <LinearGradient
        style={{
          height: 320,
          marginHorizontal: 20,
          alignSelf: 'stretch',

          alignItems: 'center',
          justifyContent: 'space-between',
        }}
        start={{ x: 0.4, y: 0.6 }} end={{ x: 1.0, y: 0.8 }}
        locations={[0, 0.85]}
        colors={[colors.dodgerBlue, 'rgb(100,199,255)']}
      >
        <StyledView>
          <TouchableOpacity
            activeOpacity={0.5}
            // onPress={goToProfile}
          >
            {
              user.photoURL
                ? <Image style={{
                  alignSelf: 'center',
                }} source={user.photoURL}/>
                : <Icon5 name="meh" size={80} color={colors.dusk} light/>
            }
          </TouchableOpacity>
          <StyledTextDisplayName>{user.displayName}</StyledTextDisplayName>
          <StyledTextStatusMsg>{user.statusMsg}</StyledTextStatusMsg>
        </StyledView>
        {
          isFriendAdded
            ? <StyledTextFriendAdded>{getString('FRIEND_ADDED')}</StyledTextFriendAdded>
            : isFriendAlreadyAdded
              ? <StyledTextFriendAlreadyAdded>{getString('FRIEND_ALREADY_ADDED')}</StyledTextFriendAlreadyAdded>
              : null
        }
        <StyledViewBtns>
          <TouchableOpacity
            testID='btn-add-or-delete'
            activeOpacity={0.5}
            onPress={showAddBtn ? addFriend : deleteFriend}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{
                showAddBtn ? getString('ADD_FRIEND') : getString('DELETE_FRIEND')
              }</StyledTextBtn>
            </View>
          </TouchableOpacity>
          <StyledViewBtnDivider />
          <TouchableOpacity
            testID='btn-chat'
            activeOpacity={0.5}
            onPress={props.onChatPressed}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{getString('CHAT')}</StyledTextBtn>
            </View>
          </TouchableOpacity>
        </StyledViewBtns>
      </LinearGradient>
    </Modal>
  );
}

export default forwardRef(Shared);
