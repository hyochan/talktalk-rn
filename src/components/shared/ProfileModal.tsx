import LinearGradient from 'react-native-linear-gradient';
import React, {
  Component,
  useState,
  useImperativeHandle,
  forwardRef,
  Ref,
  RefObject,
} from 'react';
import {
  ActivityIndicator,
  TouchableOpacity,
  StyleSheet,
  Image,
  Text,
  View,
  ViewStyle,
  ImageStyle,
  TextStyle,
  ImageSourcePropType,
} from 'react-native';
import Modal from 'react-native-modalbox';
import Icon5 from 'react-native-vector-icons/FontAwesome5';
import styled, { ThemeProps, DefaultTheme, withTheme } from 'styled-components/native';
import { User } from '../../models/User';
import { IC_MASK } from '../../utils/Icons';
import { getString } from '../../../STRINGS';

const StyledView = styled.View`
  margin-top: 40px;
`;

const StyledViewBtns = styled.View`
  height: 80px;
  align-self: stretch;
  background-color: white;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledViewBtnDivider = styled.View`
  width: 1px;
  height: 80px;
  background-color: ${({ theme: { colors: { paleGray }}}) => paleGray};
`;

const StyledTextDisplayName = styled.Text`
  font-size: 24px;
  color: white;
  font-weight: bold;
  margin-top: 32px;
  align-self: center;
`;

const StyledTextStatusMsg = styled.Text`
  font-size: 12px;
  color: white;
  margin-top: 8px;
  align-self: center;
`;

const StyledTextBtn = styled.Text`
  color: ${({ theme: { colors: { dodgerBlue }}}) => dodgerBlue};
  font-size: 16px;
`;

const StyledTextFriendAdded = styled.Text`
  color: white;
  font-size: 12px;
  background-color: ${({ theme: { colors: { dusk }}}) => dusk};
  padding: 4;px
`;

const StyledTextFriendAlreadyAdded = styled.Text`
  color: red;
  font-size: 12px;
  background-color: ${({ theme: { colors: { cloudyBlue }}}) => cloudyBlue};
  padding: 4px;
`;

interface IProps extends ThemeProps<DefaultTheme> {
  testID?: string;
  ref?: any;
  onChatPressed?: () => void;
}

interface IRef {
  open: () => void;
  close: () => void;
  setUser: (user: User) => void;
  showAddBtn: (show: boolean) => void;
}

interface IStyles {
  wrapper: ViewStyle;
  viewBtn: ViewStyle;
}

const styles: IStyles = {
  wrapper: {
    backgroundColor: 'transparent',
    alignSelf: 'stretch',
    height: 320,
    width: '90%',
    alignContent: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  viewBtn: {
    width: '50%',
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
};

const Shared = forwardRef<IRef, IProps>((props, ref) => {
  let modal: any;
  const [showAddBtn, setShowAddBtn] = useState(true);
  const [isAdding, setIsAdding] = useState(false);
  const [isFriendAdded, setIsFriendAdded] = useState(false);
  const [isFriendAlreadyAdded, setIsFriendAlreadyAdded] = useState(false);
  const [user, setUser] = useState(new User());

  const open = () => {
    setIsFriendAdded(false);
    setIsFriendAlreadyAdded(false);
    if (modal) {
      modal.open();
    }
  };

  const close = () => {
    if (modal) {
      modal.close();
    }
  };

  const addFriend = () => {

  };

  const deleteFriend = () => {
    if (modal) {
      modal.close();
    }
  };

  useImperativeHandle(ref, () => ({
    open,
    close,
    setUser: (newUser: User) => {
      setUser(newUser);
    },
    showAddBtn: (flag: boolean) => {
      setShowAddBtn(flag);
    },
  }));
  const { 
    photoURL,
    displayName,
    statusMsg
  } = user;
  const {
    theme: {
      colors: {
        dusk,
        dodgerBlue
      }
    }
  } = props;
  const imageURL = typeof photoURL === 'string' ? { uri: photoURL } : photoURL;
  return (
    <Modal
      ref={(v: any) => modal = v}
      backdropOpacity={0.075}
      entry={'top'}
      position={'center'}
      style={styles.wrapper}
    >
      <LinearGradient
        style={{
          height: 320,
          marginHorizontal: 20,
          alignSelf: 'stretch',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}
        start={{ x: 0.4, y: 0.6 }} end={{ x: 1.0, y: 0.8 }}
        locations={[0, 0.85]}
        colors={[dodgerBlue, 'rgb(100,199,255)']}
      >
        <StyledView>
          <TouchableOpacity
            activeOpacity={0.5}
            // onPress={goToProfile}
          >
            {photoURL
              ? <Image 
                  style={{ alignSelf: 'center' }} 
                  source={imageURL} 
                />
              : <Icon5 name='meh' size={80} color={dusk} light/>
            }
          </TouchableOpacity>
          <StyledTextDisplayName>{displayName}</StyledTextDisplayName>
          <StyledTextStatusMsg>{statusMsg}</StyledTextStatusMsg>
        </StyledView>
        {isFriendAdded
          ? <StyledTextFriendAdded>{getString('FRIEND_ADDED')}</StyledTextFriendAdded>
          : isFriendAlreadyAdded
            ? <StyledTextFriendAlreadyAdded>{getString('FRIEND_ALREADY_ADDED')}</StyledTextFriendAlreadyAdded>
            : null
        }
        <StyledViewBtns>
          <TouchableOpacity
            testID='btn-add-or-delete'
            activeOpacity={0.5}
            onPress={showAddBtn ? addFriend : deleteFriend}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{
                showAddBtn ? getString('ADD_FRIEND') : getString('DELETE_FRIEND')
              }</StyledTextBtn>
            </View>
          </TouchableOpacity>
          <StyledViewBtnDivider />
          <TouchableOpacity
            testID='btn-chat'
            activeOpacity={0.5}
            onPress={props.onChatPressed}
            style={styles.viewBtn}
          >
            <View style={styles.viewBtn}>
              <StyledTextBtn>{getString('CHAT')}</StyledTextBtn>
            </View>
          </TouchableOpacity>
        </StyledViewBtns>
      </LinearGradient>
    </Modal>
  );
});

export default withTheme(Shared);
