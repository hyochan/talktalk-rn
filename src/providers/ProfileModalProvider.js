import React, { createContext } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import { User as Friend } from '../models/User';
import ProfileModal from '../components/shared/ProfileModal';
import { ProfileModalContext } from './ProfileModalContext';

const ProfileModalConsumer = ProfileModalContext.Consumer;

class ProfileModalProvider extends React.Component {
  static modal: ProfileModal;

  constructor(props) {
    super(props);
    this.state = {
      user: null,
    };
  }

  actions = {
    setModal: (v) => {
      this.modal = v;
    },
    showModal: (user: Friend, deleteMode?: boolean) => {
      this.setState({ user }, () => {
        this.modal.setUser(user);
        if (this.modal) {
          this.modal.showAddBtn(!deleteMode);
          this.modal.open();
        }
      });
    },
    onChatPressed: (navigation) => {
      if (this.modal) {
        this.modal.close();
      }
      navigation.navigate('Chat');
    }
  };

  render() {
    const { state, actions, modal } = this;
    const value = { state, actions, modal };
    return (
      <ProfileModalContext.Provider value={value}>
        {this.props.children}
        <ProfileModal
          id="modal"
          ref={(v) => {
            this.modal = v;
          }}
          onChatPressed={ () => this.actions.onChatPressed(this.props.navigation) }
        />
      </ProfileModalContext.Provider>
    );
  }
}

export { ProfileModalConsumer, ProfileModalProvider };
