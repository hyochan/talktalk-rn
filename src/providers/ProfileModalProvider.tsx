import React, { Component, useReducer, useState, useRef } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import { NavigationScreenProp, NavigationStateRoute } from 'react-navigation';
import { User as Friend } from '../models/User';
import ProfileModal from '../components/shared/ProfileModal';
import { ProfileModalContext } from '../contexts/ProfileModalContext';

const ProfileModalConsumer = ProfileModalContext.Consumer;

interface IProps {
  navigation: NavigationScreenProp<NavigationStateRoute<any>>;
  children?: any;
}

interface IState {
  user: Friend;
  deleteMode: boolean;
}

interface IAction {
  type: 'show-modal';
  payload: IState;
}

const initialState: IState = {
  user: {
    uid: '',
    displayName: '',
    photoURL: '',
    statusMsg: '',
  },
  deleteMode: false,
};

let modal: any;

const reducer = (state: IState, action: IAction) => {
  switch (action.type) {
    case 'show-modal':
      if (modal && modal.current) {
        modal.current.setUser(action.payload.user);
        modal.current.showAddBtn(!action.payload.deleteMode);
        modal.current.open();
      }
      return {
        ...state,
        user: action.payload.user,
        deleteMode: !action.payload.deleteMode,
      };
    default:
      return state;
  }
};

function ProfileModalProvider(props: IProps) {
  modal = useRef(null);
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = { state, dispatch };

  return (
    <ProfileModalContext.Provider value={value}>
      {props.children}
      <ProfileModal
        testID='modal'
        ref={modal}
        onChatPressed={() => {
          if (modal && modal.current) {
            modal.current.close();
          }
          props.navigation.navigate('Chat');
        }}
      />
    </ProfileModalContext.Provider>
  );
}

export { ProfileModalConsumer, ProfileModalContext, ProfileModalProvider };
