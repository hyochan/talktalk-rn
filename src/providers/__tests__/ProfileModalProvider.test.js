import 'react-native';
import * as React from 'react';
import { ProfileModalProvider } from '../ProfileModalProvider';
import ProfileModal from '../../components/shared/ProfileModal';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow, render, mount } from 'enzyme';

describe('[ProfileModalProvider] rendering test', () => {
  const wrapper = shallow(<ProfileModalProvider/>);
  it('[ProfileModalProvider] renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('[ProfileModalProvider] interaction', () => {
  let props;
  let providerWrapper;
  let modalWrapper;
  const friend = {
    uid: '2',
    displayName: 'geoseong',
    photoURL: null,
    statusMsg: 'healthy',
    isOnline: '',
    friends: '',
    Chatrooms: '',
    created: '',
    updated: '',
  };

  beforeEach(() => {
    props = {
      navigation: {
        navigate: jest.fn(),
      },
      state: {
        user: null,
      },
      actions: {
        setModal: jest.fn(),
        showModal: jest.fn(),
        onChatPressed: jest.fn(),
      }
    };
    providerWrapper = shallow(<ProfileModalProvider {...props}/>);
    modalWrapper = providerWrapper.find('#modal');
  });
  it('check actions', () => {
    modalWrapper.props().onChatPressed();
    expect(props.navigation.navigate).toHaveBeenCalled();
  });
});
