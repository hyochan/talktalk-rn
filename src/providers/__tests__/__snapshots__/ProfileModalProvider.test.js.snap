// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProfileModalProvider] rendering test [ProfileModalProvider] renders as expected 1`] = `
<ContextProvider
  value={
    Object {
      "actions": Object {
        "onChatPressed": [Function],
        "setModal": [Function],
        "showModal": [Function],
      },
      "modal": undefined,
      "state": Object {
        "user": null,
      },
    }
  }
>
  <Shared
    id="modal"
    onChatPressed={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "height": 320,
        "justifyContent": "space-between",
        "marginHorizontal": 20,
      }
    }
  />
</ContextProvider>
`;
